### ISSUE TITLE
Update a counter record when add or remove a table

### ISSUE THREAD

**Issue by drmingdrmer** [datafuselabs/databend#5677]
> Would a `list_id_value` or `list_keys` call different from common read calls? 

`list_id_value` can be implemented with a `prefix-list` RPC to metasrv.
`list_keys` can not be expressed with KVApi yet.

> Moreover, would it be ok for meta to provide some api for stats grouped by prefix (such as tenant id) ?

metasrv does not know what the delimiter is for an application, although traditionally it is a slash `/`.
And maintaining the counts of every prefix will result in unnecessary write-amplification.

Thus IMHO, `count by the prefix` should be implemented on the application side:

When creating/dropping/undropping/renaming a table, also update a counter record.

_Originally posted by @drmingdrmer in https://github.com/datafuselabs/databend/issues/5646#issuecomment-1140803314_

**Comment by drmingdrmer** [comment]
- [ ] TODO: @everpcpc Does the counter count the dropped but not cleaned tables? I.e., tables that can be `undropped`.

**Comment by drmingdrmer** [comment]
@RinChanNOWWW

**Comment by RinChanNOWWW** [comment]
Where should I place the struct `TableCounter` ?

**Comment by drmingdrmer** [comment]
> Where should I place the struct `TableCounter` ?

`common/meta/app/src/schema/table.rs` would be a good place for it

**Comment by everpcpc** [comment]
> Does the counter count the dropped but not cleaned tables? I.e., tables that can be `undropped`.

I think either would be ok.

**Comment by RinChanNOWWW** [comment]
Do applications get count of tables by `SchemaApi.list_tables` and compute the size of the Vec? If so, should we separate an API like `count_tables`? 

I come out with two ideas:

1. Separate a independent API `SchemaApi.count_tables`, in this case,  `TableCounter` only needs to record the number of tables.
2. Continue to use `SchemaApi.list_tables`, in this case, we will not record the counts but the table_ids.

Do you think which one is better?

**Comment by drmingdrmer** [comment]
count_table sounds good.

BTW this issue is meant to get the count of all tables belonging to a tenant without listing tables.

**Comment by RinChanNOWWW** [comment]
> count_table sounds good.
> 
> BTW this issue is meant to get the count of all tables belonging to a tenant without listing tables.

I think so. So I'm wondering what is the current implementation.

**Comment by drmingdrmer** [comment]
> I think so. So I'm wondering what is the current implementation.

You can find it in the referenced PR at the bottom of the issue body

**Comment by RinChanNOWWW** [comment]
Should the operation of updating count of table be under transaction?

Codes in `create_table`: 

```rust
let txn_req = TxnRequest {
condition: vec![
  // db has not to change, i.e., no new table is created.
  // Renaming db is OK and does not affect the seq of db_meta.
  txn_cond_seq(&DatabaseId { db_id }, Eq, db_meta_seq)?,
  // no other table with the same name is inserted.
  txn_cond_seq(&dbid_tbname, Eq, 0)?,
  // no other table id with the same name is append.
  txn_cond_seq(&dbid_tbname_idlist, Eq, tb_id_list_seq)?,
],
if_then: vec![
  // Changing a table in a db has to update the seq of db_meta,
  // to block the batch-delete-tables when deleting a db.
  // TODO: test this when old metasrv is replaced with kv-txn based SchemaApi.
  txn_op_put(&DatabaseId { db_id }, serialize_struct(&db_meta)?)?, // (db_id) -> db_meta
  txn_op_put(&dbid_tbname, serialize_id(table_id)?)?, // (tenant, db_id, tb_name) -> tb_id
  txn_op_put(&tbid, serialize_struct(&req.table_meta)?)?, // (tenant, db_id, tb_id) -> tb_meta
  txn_op_put(&dbid_tbname_idlist, serialize_struct(&tb_id_list)?)?, // _fd_table_id_list/db_id/table_name -> tb_id_list
],
else_then: vec![],
};

let (succ, _responses) = send_txn(self, txn_req).await?;
```

If so, how do we guarantee the atomicity of getting the old count and putting the new count?

**Comment by drmingdrmer** [comment]
Every time a value is updated, the `seq` for it will be increased too.
By asserting the `seq` being the same to ensure the value you read is not changed., e.g: `txn_cond_seq(&DatabaseId { db_id }, Eq, db_meta_seq)?` ensures the `db_meta` is not changed.

**Comment by RinChanNOWWW** [comment]
Oh, I know. We can use the seq number, just like the other codes.

**Comment by RinChanNOWWW** [comment]
I think I can reuse the function `get_id_value`, `serialize_id` and `deserialize_id` for the count value? But it will influence the readability.

**Comment by drmingdrmer** [comment]
> I think I can reuse the function `get_id_value`, `serialize_id` and `deserialize_id` for the count value? But it will influence the readability.

Right. you can just rename these functions to make them a little bit general.
Maybe `get_u64_xxx`? I am not quite sure yet.

**Comment by RinChanNOWWW** [comment]
Should we consider the backward compatibility? 

If we find the count is 0, we compute the count in the old way first (iterate all the dbs and list tables). and put the value lazily?

OR

Check the table-count key space for all tenants when metasrv is launched?

**Comment by drmingdrmer** [comment]
> Should we consider the backward compatibility?
> 
> If we find the count is 0, we compute the count in the old way first (iterate all the dbs and list tables). and put the value lazily?
> 
> OR
> 
> Check the table-count key space for all tenants when metasrv is launched?

What's the best way you think? :)

I do not know if it's appropriate for me to give you every detailed decision.

**Comment by RinChanNOWWW** [comment]
I'm implementing the first way now. Let me finish the whole task first.
